pallene/reaching_definitions.lua

```diff
@@ -0,0 +1,497 @@
+        return self:new_edge(loop, ctx)
+    elseif tag == "ir.Cmd.Loop" then
+        Stack.push(ctx.current_loop, {})
+        --print('create new loop',#ctx.current_loop)
+        local newctx = self:build(cmd.body,ctx)
+        -- at the end of the loop, gets all indexes at the top 
+        -- and makes them point to the next (TODO)
+        -- will have to change the edge logic on break
+        Stack.pop(newctx.current_loop)
+        return newctx
+    elseif tag == "ir.Cmd.If" then
+        -- condition is an ir.Value and 
+        -- will be extracted to an assignment instruction beforehand
+        local then_ = cmd.then_ and self:build(cmd.then_,ctx) or {}
+        local else_ = cmd.else_ and self:build(cmd.else_,ctx) or {}
+        return CFG.ctx_union(then_,else_)
+    elseif tag == "ir.Cmd.For" then
+        -- start,limit,step are ir.Value and 
+        -- will be extracted to assignments beforehand
+        Stack.push(ctx.current_loop, ctx.current_node)
+        local newctx = self:build(cmd.body,ctx)
+        Stack.pop(newctx.current_loop)
+        return newctx
+    elseif string.match(tag, "^ir%.Cmd%.") then
+        return self:new_node(cmd,ctx)
+    else
+        return ctx
+    end
+end
+-- returns a copy of the context
+-- a context has:
+--     current_node : set<index>,  
+--     current_loop : stack<index>
+function CFG.new_ctx(ctx)
+    local new = {}
+    new.current_node = {}
+    new.current_loop = {}
+    if ctx then
+        Set.copy(new.current_node,ctx.current_node)
+        Stack.copy(new.current_loop,ctx.current_loop)
+    end
+    return new
+end
+--performs union on two contexts, returning a new one
+function CFG.ctx_union(ctx1,ctx2)
+    local new = {}
+    new.current_node = Set.union(ctx1.current_node,ctx2.current_node)
+    --TODO: assert that the loop ended? 
+    new.current_loop = Set.union(ctx1.current_loop,ctx2.current_loop)
+    return new
+end
+-- creates a new node, should only be called to non-control commands.
+-- returns an updated context
+function CFG:new_node(cmd,ctx)
+    table.insert(self,{
+        cmd = cmd, --TODO should be a list of commands
+        to={},
+        from={}
+    })
+    local new_node_id = #self
+    local new_node = self[new_node_id]
+    Set.copy(new_node.from, ctx.current_node)
+    for _,from in ipairs(new_node.from) do
+            table.insert(self[from].to,new_node_id)
+    end
+    local newctx = CFG.new_ctx(ctx)
+    newctx.current_node = {new_node_id}
+    return newctx
+end
+-- creates a new edge from each of the current nodes to each of the target nodes
+function CFG:new_edge(target,ctx)
+    for _,from in ipairs(ctx.current_node) do
+        for _,to in ipairs(target) do
+            table.insert(self[from].to,to)
+            table.insert(self[to].from,from)
+        end
+    end
+    return ctx
+end
+
+-- traverses the CFG applying the array of functions fs with the input to each node.
+function CFG:traverse(index,fs,input,visited)
```

Comment:
1. What order is the traversal? That should be clearly specified in the comment.
2. A single `f` instead of `fs` would work just as well.
3. We don't need a function parameter because Lua has closures. That is, instead of `traverse(cfg, f, x)` the caller can  always write `traverse(cfg, function(cfg, i) f(cfg, i, x) end`)
4. Visited is only use internally so it should not be a parameter. It would be better to define the visited array as a local variable to the traverse method and to have nested function that does the actual recursion. I like using the name `go` for the internal function. See ir.iter for an example.

```
function function traverse(cfg, root_index)
    local visited = {}
    local  go(i)
        -- ...
        visited[i] = true
    end
```

Reply:
ok